### PR TITLE
Backport 2.x: Fix alert raised for invalid fragment length

### PR DESCRIPTION
**Pull request by daverodgman**
Backport of the bugfix part of #4724.

### CODE REVIEW

File: ChangeLog.d/fix_tls_alert_codes.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Bugfix
+   * Fix the alert raised when a client requests an invalid
+   * fragment length, as per RFC6066 section 4. We now alert with
+   * MBEDTLS_SSL_ALERT_MSG_ILLEGAL_PARAMETER. Similarly, raise
+   * MBEDTLS_SSL_ALERT_MSG_DECRYPT_ERROR for an invalid finished
+   * message, as per RFC5247 section 7.2.2.
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/fix_tls_alert_codes.txt:6]: See my comments on the main PR.

File: ChangeLog.d/fix_tls_alert_codes.txt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Bugfix
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/fix_tls_alert_codes.txt:1]: What about the unsupported_ext cases?

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

File: include/mbedtls/ssl.h [hunk 3]
```diff
@@ -133,6 +133,7 @@
 #define MBEDTLS_ERR_SSL_VERSION_MISMATCH                  -0x5F00  /**< An operation failed due to an unexpected version or configuration. */
 #define MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS                -0x7000  /**< A cryptographic operation is in progress. Try again later. */
 #define MBEDTLS_ERR_SSL_BAD_CONFIG                        -0x5E80  /**< Invalid value in SSL config */
+#define MBEDTLS_ERR_SSL_UNSUPPORTED_EXTENSION             -0x5E00  /**< Client received an extended server hello containing an unsupported extension */
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h:136]: We shouldn't add a new error code in 2.x. Please change the alert only.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Ok for the alert changes, but the error codes must not change.

**Review by ronald-cron-arm — DISMISSED** [review]
This PR backports the alert message changes flagged as bug fix in #4724 and only this. LGTM.

**Review by hanno-arm — DISMISSED** [review]

**Review by gilles-peskine-arm — APPROVED** [review]

File: tests/ssl-opt.sh [hunk 4]
```diff
@@ -5957,7 +5957,7 @@ run_test    "ECJPAKE: client not configured" \
             -S "ciphersuite mismatch: ecjpake not configured" \
             -S "server hello, ecjpake kkpp extension" \
             -C "found ecjpake_kkpp extension" \
-            -S "None of the common ciphersuites is usable"
```
Comment (gilles-peskine-arm) [inline tests/ssl-opt.sh:5960]: No, this hasn't changed in 2.x or 2.16

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by ronald-cron-arm — APPROVED** [review]

**Pull request merged**